execv: fix execv returns 0 and doesn't set errno

Ahish9009 opened the pull request:
Closes #1335:
Behaves like execv in posix: returns -1 on failure and sets errno to the value of error that occurred. Seems to have fixed #1334, but does not seem to have helped #1702.

Ahish9009 commented:
Not sure what's wrong in the last check. Seems to have stopped while pinging 10.0.2.10.
Don't see why my changes should have affected it.

anton-bondarev commented:
> Not sure what's wrong in the last check. Seems to have stopped while pinging 10.0.2.10.
> Don't see why my changes should have affected it.

Don't worry about this. It's not your fault. We restarted CI and it passed succeed

anton-bondarev commented:
@Ahish9009 Hi, Thanks
I think You on the right way.
But Why do you want to differ in two situations

1. When cmd_exec return with errors
2. When cmd has not been found

I supposed You can just return ecode from exec_call and then return it in execv

Ahish9009 commented:
I had a look at the ```cmd_exec``` code and realised that the check for whether the command exists is happening already, but is returning EINVAL instead of ENOENT, which I fixed.
Further, not testing whether the cmd exists in ```exec_call()``` caused the issue of passing a NULL pointer to ```cmd2mod()```, which is also fixed now.
Please have a look at the changes.

anton-bondarev commented on src/cmds/sys/env.c:
```diff
@@ -68,8 +68,7 @@ int main(int argc, char *argv[]) {
 	/* difference from posix: execv return 0 and does not set errno on errors */
 	if (optind < argc) {
 		execv(argv[optind], argv+optind);
-		printf("env: execv: No such command '%s'\n", argv[optind]);
```
I think It's better to keep printf but the message change to "env: execv: command '%s' failed with error code %d"

Ahish9009 replied:
Okay, will make this change, but that'll mean I'll be printing errno(since ecode won't be returned from execv, as POSIX requires execv to return -1 on errors). 
So if I'm printing errno, isn't it better to use perror()?

anton-bondarev replied:
Ok, just remove.
Thanks

anton-bondarev commented on src/include/framework/cmd/types.h:
```diff
@@ -37,6 +37,7 @@ struct cmd_mod {
 };
 
 static inline const struct mod *cmd2mod(const struct cmd *cmd) {
+	if (!cmd) return NULL;
```
I think You should check this before calling cmd2mod

Ahish9009 replied:
I did this so that I don't have to perform this check in the exec_call() function.

Ahish9009 replied:
Reverted

anton-bondarev commented on src/compat/posix/proc/exec.c:
```diff
@@ -55,18 +57,10 @@ int exec_call(void) {
 		const struct cmd *cmd;
 
 		cmd = cmd_lookup(cmd_name);
-
```
Why you remove this?
Revert please

Ahish9009 replied:
I removed it because cmd_exec checks if the cmd exists anyways, so found that this check is redundant. Should I revert back though?

anton-bondarev replied:
Yes, I see. But I don't understand why did you move checking from here? Why did you place it in cmd2mod(). why did you move it at all?

Ahish9009 replied:
My aim was to remove setting errno to ENOENT in the exec_call() function, so I removed the check in exec_call() and simply used the check already present in cmd_exec(). 
On running however, as cmd was NULL, cmd2mod() caused errors as it was trying to perform its operation on a NULL pointer. So I included the check there too.
So should I put the check back here and remove the check in cmd2mod()?

anton-bondarev replied:
> My aim was to remove setting ecode and errno to ENOENT
Why? :)

I thought your aim is "execv returns 0 and doesn't set errno" :)

Ahish9009 replied:
> But Why do you want to differ in two situations
>
> When cmd_exec return with errors
> When cmd has not been found
> I supposed You can just return ecode from exec_call and then return it in execv

I thought when you said this in the earlier comment, you wanted me to set all errors in the cmd_exec() function itself, my bad. Should I revert it back to the way it was?

anton-bondarev replied:
> Should I revert it back to the way it was?
Yes please.
I'm sorry that uncleanly put it

Ahish9009 replied:
Done :)

anton-bondarev commented on src/framework/cmd/core.c:
```diff
@@ -23,8 +23,10 @@ extern void getopt_init(void);
 
-	if (!cmd)
-		return -EINVAL;
+	if (!cmd) {
```
the same question.
Why did you change src/framework/cmd/core.c?
why you could not do it in src/compat/posix/proc/exec.c?

Ahish9009 replied:
It can be done in either place, but I thought it'd be better to do it in the cmd_exec() function as the error is during execution of the cmd, but I can revert back if needed, not a problem :)

Ahish9009 replied:
Reverted

anton-bondarev commented:
@Ahish9009 Thank you. I've approved PR.
Please rebase and squash commits

Ahish9009 commented:
> @Ahish9009 Thank you. I've approved PR.
> Please rebase and squash commits

Done
